gensim/similarity_learning/custom_losses.py

```diff
@@ -0,0 +1,10 @@
+from keras import backend as K
+from keras.layers import Lambda
+
+
+def rank_hinge_loss(y_true, y_pred):
+    margin = 1
+    y_pos = Lambda(lambda a: a[::2, :], output_shape=(1,))(y_pred)
+    y_neg = Lambda(lambda a: a[1::2, :], output_shape=(1,))(y_pred)
+    loss = K.maximum(0., margin + y_neg - y_pos)

```

Comment:
Check the hinge-loss definition, needs correction 
https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/python/ops/losses/losses_impl.py

Reply:
https://github.com/keras-team/keras/blob/3eab6103e0069d261c3e3586c12d5aaf0a331429/keras/losses.py#L38

I updated it now to have the  default keras hinge loss

The validation results are posted after every epoch.
The test result is posted at the end as a dict.

The validation metrics don't change at all. There might be some bug somewhere.
Here's the log:
```
2018-06-09 19:42:47,239 : INFO : Starting Vocab Build
2018-06-09 19:42:47,756 : INFO : Vocab Build Complete
2018-06-09 19:42:47,756 : INFO : Vocab Size is 30772
2018-06-09 19:42:47,757 : INFO : Building embedding index using pretrained word embeddings
2018-06-09 19:42:55,247 : INFO : The embeddings_index built from the given file has 400000 words of 50 dimensions
2018-06-09 19:42:55,248 : INFO : Embedding Matrix for Embedding Layer has shape (30773, 50)
2018-06-09 19:42:55,301 : INFO : There are 1685 words not in the embeddings. Setting them to zero
2018-06-09 19:42:55,301 : INFO : Adding additional dimensions from the embedding file to embedding matrix
2018-06-09 19:42:56,307 : INFO : Embedding Matrix now has shape (401688, 50)
2018-06-09 19:42:56,307 : INFO : Pad word has been set to index 401685
2018-06-09 19:42:56,308 : INFO : Embedding index build complete
2018-06-09 19:42:58,175 : INFO : 0 queries were filtered
2018-06-09 19:42:58,178 : INFO : There are a total of 872 queries
2018-06-09 19:42:58,189 : INFO : There are 8990 pairs to train on. So, a total of 17980 datapoints
2018-06-09 19:42:58,211 : INFO : Getting List Vocab from given vocab
2018-06-09 19:42:58,344 : INFO : Vocab Transfer complete. Method used for OOV is ignore.
2018-06-09 19:42:58,345 : INFO : There are a total of 254 unkown words
2018-06-09 19:42:58,391 : INFO : Getting List Vocab from given vocab
2018-06-09 19:42:58,612 : INFO : Vocab Transfer complete. Method used for OOV is ignore.
2018-06-09 19:42:58,612 : INFO : There are a total of 674 unkown words

2018-06-09 19:42:58.867334: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2018-06-09 19:42:58.868072: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2018-06-09 19:42:58.870084: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2018-06-09 19:42:58.872295: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2018-06-09 19:42:58.873052: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to
==================================================================================================
query (InputLayer)              (None, 140)          0
__________________________________________________________________________________________________
doc (InputLayer)                (None, 140)          0
__________________________________________________________________________________________________
embedding_1 (Embedding)         (None, 140, 50)      20084400    query[0][0]
                                                                 doc[0][0]
__________________________________________________________________________________________________
mm_q_embed_DOT_d_embed (Dot)    (None, 140, 140)     0           embedding_1[0][0]
                                                                 embedding_1[1][0]
__________________________________________________________________________________________________
mm_k_topk_mm (Lambda)           (None, 140, 20)      0           mm_q_embed_DOT_d_embed[0][0]
__________________________________________________________________________________________________
mm_k_Dense_5_mm_k (Dense)       (None, 140, 5)       105         mm_k_topk_mm[0][0]
__________________________________________________________________________________________________
mm_k_Dense_1_mm_k (Dense)       (None, 140, 1)       6           mm_k_Dense_5_mm_k[0][0]
__________________________________________________________________________________________________
w_g_Dense_1_q_embed (Dense)     (None, 140, 1)       51          embedding_1[0][0]
__________________________________________________________________________________________________
mm_k_dropout_Dropout_mm_k (Drop (None, 140, 1)       0           mm_k_Dense_1_mm_k[0][0]
__________________________________________________________________________________________________
g_Softmax_w_g (Lambda)          (None, 140)          0           w_g_Dense_1_q_embed[0][0]
__________________________________________________________________________________________________
mm_reshape_Reshape_maxlen_mm_k_ (None, 140)          0           mm_k_dropout_Dropout_mm_k[0][0]
__________________________________________________________________________________________________
g_Reshape_maxlen_w_g (Reshape)  (None, 140)          0           g_Softmax_w_g[0][0]
__________________________________________________________________________________________________
mean_mm_reshape_DOT_g (Dot)     (None, 1)            0           mm_reshape_Reshape_maxlen_mm_k_dr
                                                                 g_Reshape_maxlen_w_g[0][0]
__________________________________________________________________________________________________
out_Reshape_mean (Reshape)      (None, 1)            0           mean_mm_reshape_DOT_g[0][0]
==================================================================================================
Total params: 20,084,562
Trainable params: 162
Non-trainable params: 20,084,400
__________________________________________________________________________________________________
Epoch 1/10
17980/17980 [==============================] - 11s 626us/step - loss: 0.5000 - acc: 0.0000e+00
MAP:  0.4272458190087743
nDCG@ 1 :  0.24603174603174602
nDCG@ 3 :  0.4039413159197722
nDCG@ 5 :  0.48513949880292767
nDCG@ 10 :  0.5501837173391042
nDCG@ 20 :  0.5706403081981417
Epoch 2/10
17980/17980 [==============================] - 12s 642us/step - loss: 0.5000 - acc: 0.0000e+00
MAP:  0.4272458190087743
nDCG@ 1 :  0.24603174603174602
nDCG@ 3 :  0.4039413159197722
nDCG@ 5 :  0.48513949880292767
nDCG@ 10 :  0.5501837173391042
nDCG@ 20 :  0.5706403081981417
Epoch 3/10
17980/17980 [==============================] - 11s 630us/step - loss: 0.5000 - acc: 0.0000e+00
MAP:  0.4272458190087743
nDCG@ 1 :  0.24603174603174602
nDCG@ 3 :  0.4039413159197722
nDCG@ 5 :  0.48513949880292767
nDCG@ 10 :  0.5501837173391042
nDCG@ 20 :  0.5706403081981417
Epoch 4/10
17980/17980 [==============================] - 11s 621us/step - loss: 0.5000 - acc: 0.0000e+00
MAP:  0.4272458190087743
nDCG@ 1 :  0.24603174603174602
nDCG@ 3 :  0.4039413159197722
nDCG@ 5 :  0.48513949880292767
nDCG@ 10 :  0.5501837173391042
nDCG@ 20 :  0.5706403081981417
Epoch 5/10
17980/17980 [==============================] - 11s 594us/step - loss: 0.5000 - acc: 0.0000e+00
MAP:  0.4272458190087743
nDCG@ 1 :  0.24603174603174602
nDCG@ 3 :  0.4039413159197722
nDCG@ 5 :  0.48513949880292767
nDCG@ 10 :  0.5501837173391042
nDCG@ 20 :  0.5706403081981417
Epoch 6/10
17980/17980 [==============================] - 11s 613us/step - loss: 0.5000 - acc: 0.0000e+00
MAP:  0.4272458190087743
nDCG@ 1 :  0.24603174603174602
nDCG@ 3 :  0.4039413159197722
nDCG@ 5 :  0.48513949880292767
nDCG@ 10 :  0.5501837173391042
nDCG@ 20 :  0.5706403081981417
Epoch 7/10
17980/17980 [==============================] - 11s 620us/step - loss: 0.5000 - acc: 0.0000e+00
MAP:  0.4272458190087743
nDCG@ 1 :  0.24603174603174602
nDCG@ 3 :  0.4039413159197722
nDCG@ 5 :  0.48513949880292767
nDCG@ 10 :  0.5501837173391042
nDCG@ 20 :  0.5706403081981417
Epoch 8/10
17980/17980 [==============================] - 11s 611us/step - loss: 0.5000 - acc: 0.0000e+00
MAP:  0.4272458190087743
nDCG@ 1 :  0.24603174603174602
nDCG@ 3 :  0.4039413159197722
nDCG@ 5 :  0.48513949880292767
nDCG@ 10 :  0.5501837173391042
nDCG@ 20 :  0.5706403081981417
Epoch 9/10
17980/17980 [==============================] - 46s 3ms/step - loss: 0.5000 - acc: 0.0000e+00
MAP:  0.4272458190087743
nDCG@ 1 :  0.24603174603174602
nDCG@ 3 :  0.4039413159197722
nDCG@ 5 :  0.48513949880292767
nDCG@ 10 :  0.5501837173391042
nDCG@ 20 :  0.5706403081981417
Epoch 10/10
17980/17980 [==============================] - 10s 584us/step - loss: 0.5000 - acc: 0.0000e+00
MAP:  0.4272458190087743
nDCG@ 1 :  0.24603174603174602
nDCG@ 3 :  0.4039413159197722
nDCG@ 5 :  0.48513949880292767
nDCG@ 10 :  0.5501837173391042
nDCG@ 20 :  0.5706403081981417
2018-06-09 19:45:45,220 : INFO : Results of evaluating on DRMM_TKS:
2018-06-09 19:45:45,221 : INFO : {'ndcg@1': 0.1646090534979424, 'method': 'DRMM_TKS', 'map': 0.3798646745991278, 'ndcg@5': 0.4557856003454745, 'ndcg@10': 0.5128600642573993, 'ndcg@20': 0.5417893700826042, 'ndcg@3': 0.37272685975711767}
2018-06-09 19:45:45,288 : INFO : Results saved in results_drmm_tks.csv
```